1443 unverified consent requests

TheAndrewJackson opened the pull request:
Closes #1443
Closes #1525

### Code Changes

* Deleted duplicated component file
* Updated privacy center to support unverified consent requests
* Updated backend to allow unverified consent requests
* Refactored unit tests to factor in the new changes
* Added new unit tests for the unverified get route

### Steps to Confirm

* [ ] Run fides with `subject_identity_verification_required` set to `false`
* [ ] Run privacy center
* [ ] Do a consent request and get redirected to the consent page
* [ ] Update preferences
* [ ] Do another consent request and see the updated preferences

### Pre-Merge Checklist

* [x] All CI Pipelines Succeeded
* Documentation Updated:
  * [ ] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  * [x] documentation issue created (tag docs-team to complete issue separately)
* [x] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [ ] Update `CHANGELOG.md`

### Description Of Changes

Update fides and the privacy center to support unverified consent requests. This will allow for easier setup and development.

https://user-images.githubusercontent.com/17103888/196696745-19308251-2524-4588-8725-e7361f94e244.mov

allisonking commented on clients/privacy-center/pages/consent.tsx:
```diff
@@ -26,10 +26,10 @@ const Consent: NextPage = () => {
       router.push("/");
     }
-  }, [consentRequestId, verificationCode, router]);
+  }, [consentRequestId, router]);
```
I don't know the expected flow of this too well, but I wonder if there is a place where we need to clear out stale values? 

or perhaps instead of local storage, we should use query parameters? though that would only work if this page and the home page are the only places that use these codes/ids. probably out of the scope of this PR tho

TheAndrewJackson replied:
There probably is a better way to do this. I think the privacy center will need a refactor to use RTK query at some in the near future. Maybe those values could live in the redux store?

allisonking commented on clients/privacy-center/components/modals/consent-request-modal/ConsentRequestModal.tsx:
```diff
@@ -21,9 +21,8 @@ export const useConsentRequestModal = () => {
 
   const successHandler = useCallback(() => {
-    setConsentRequestId(consentRequestId);
     router.push("consent");
```
a toast here might be nice. I think they do persist across route changes—if you trigger a toast on this page https://chakra-ui.com/docs/components/toast and then click on another link, the toast will still stick around)

TheAndrewJackson replied:
ohh nice! I specifically didn't use a toast because of the page change. I didn't know they persisted across pages

TheAndrewJackson replied:
I'm running into some weird issues getting the toast working. Are you okay with it being addressed in a follow up?

TheAndrewJackson replied:
The issue is react no-op errors. I'm not quite sure what's causing the no-op

allisonking replied:
yeah no problem! let's do a follow up issue since iirc there are a few places that might be nice to have a toast

allisonking reviewed (commented):
awesome python tests! just a few comments, but for the most part looks good and worked on my machine 👍

TheAndrewJackson commented on src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -137,6 +135,36 @@ def consent_request_verify(
+        )
+
+    provided_identity = _get_consent_request_and_provided_identity(
+        db=db, consent_request_id=consent_request_id, verification_code=""
```
I agree. Updated!

TheAndrewJackson commented on clients/privacy-center/components/modals/consent-request-modal/ConsentRequestForm.tsx:
```diff
@@ -74,12 +76,12 @@ const useConsentRequestForm = ({
         const data = await response.json();
```
That does make more sense. I put that check in the top level `useEffect` initially and never refactored it back out later. I'll update it so the consent id check is closer to the actual logic it effects

TheAndrewJackson replied:
The change is pushed now. If things are still a little confusing maybe I can write out the excepted flow

allisonking replied:
it looks like the changes you made were to `consent.tsx` but the function I found confusing was the `onSubmit` func in `ConsentRequestForm.tsx`. lemme know if I'm missing something about how those two are related!

TheAndrewJackson commented on src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -137,6 +135,36 @@ def consent_request_verify(
 
+@router.get(
+    CONSENT_REQUEST_PREFERENCES_WITH_ID,
+    status_code=HTTP_200_OK,
```
The `_get_consent_request_and_provided_identity` function inside this route handles raising the 400 and 404 error codes in a consent request doesn't exist.

If the consent request exists and the user has no consent preferences the `_prepare_consent_preferences` functions returns `None` for the list of preferences. I _think_ everything is handled. Does that make sense?

allisonking replied:
ah yeah what I meant here was more docs related: if you go to localhost:8080/docs, and then go to the entry for `consent`/GET, it only shows possible return statuses of 200 and 422 (not 400 or 404). here's an example of another endpoint that does it: https://github.com/ethyca/fides/blob/1443_unverified_consent_requests/src/fides/api/ctl/routes/datamap.py/#L72-L83

TheAndrewJackson commented:
@allisonking I think I've addressed all the PR feedback. Now I'm getting some test failures that seems unrelated to my code changes in the ctl test suite. I tested this locally on my machine and the tests failed on main too. I think this may be good to merge if the code looks good to you but it would be good to confirm that the tests fail for someone else too.

seanpreston commented:
It looks like the modal text says "we will email you..."

<img width="472" alt="image (3)" src="https://user-images.githubusercontent.com/1667340/196999280-77466780-1215-4e29-9aa7-638af9535857.png">

however Fides won't email the user currently. Can we correct that?

TheAndrewJackson commented:
@seanpreston good catch. Yes, I'll update that now

NevilleS commented:
@TheAndrewJackson I can try this out using `nox -s test_env` when you're ready, too. You'll notice that I disabled consent requests on `main` in a recent commit so that the privacy center would work by default.

TheAndrewJackson commented:
@NevilleS This is ready to test with `nox -s test_env` now

NevilleS commented:
@TheAndrewJackson I pulled your branch and tried testing this out and it wasn't working for me. I get a 404 error for `http://localhost:8080/api/v1/consent-request//preferences` in the Javascript console:
<img width="1677" alt="image" src="https://user-images.githubusercontent.com/1834295/197040215-e312765f-7136-47d3-af4c-2b1a929efb83.png">

Steps to reproduce:
1. `git pull 1443_unverified_consent_requests`
2. Ensure that `.fides/fides.toml` has `subject_identity_verification_required` set to `false`
3. Edit `clients/privacy-center/config/config.json` and set `"includeConsent": true`
4. Run `nox -s test_env`
5. Open privacy center at http://localhost:3001/
6. Click "Manage my consent" and provide my email
7. Nothing happens. Observe the Javascript error

Two observations here:
* It seems to be calling the API but not actually providing the email I gave it
* There is no visible error message in the UI; we should catch any server error like this and show an error toast

Let me know if you need help repro'ing this or if I've missed something.

TheAndrewJackson commented:
@NevilleS Thanks for taking a look. I followed those repo steps and everything worked for me. I thought I fixed the issue that you're seeing. Is the most recent local commit for this branch `Fix logic issue from refactor` on your machine?

An error toast would be help. Do you want it included in this PR or in a follow up ticket?

NevilleS commented:
Let's add the error toast now, or it'll just be a bug that we find later!

Meanwhile I'll double-check my repo...

NevilleS commented:
Re-tested with the latest commit and it looks good now 👍 

My little design/quality asks are:
* add error toast for server errors
* add success toast for after you save, like "Your consent preferences have been saved"

Otherwise LGTM, thanks!

NevilleS commented:
Oh, and one more thing @TheAndrewJackson - since you've done the work here to allow consent requests to be testable by default (i.e. without any secrets!), I've gone ahead and pushed a small commit to your branch to enable them in the privacy center again and give a few semi-realistic examples. That way, consent requests will be available in the test env once this merges.

Cheers!

ThomasLaPiana commented:
love to see that `test_env` usage :D

TheAndrewJackson commented:
@NevilleS The project has been updated to use toasts for everything now 🎉  The original toasts were actually a hacky workaround needed due to a config issue in the project. I removed the workaround now that everything is using actual toasts. 

Here are some examples of success toasts.

https://user-images.githubusercontent.com/17103888/197212688-7a35f08d-bc77-4520-a12a-3593a4487e5e.mov

https://user-images.githubusercontent.com/17103888/197212710-9c3664c3-ebd5-4fd9-b531-9dd6afd3951f.mov


https://user-images.githubusercontent.com/17103888/197212738-2632acd4-41f2-4eb4-865c-2c73d5388c21.mov


Users are now told if the privacy center had an issue getting the config (ex: backend isn't running)

https://user-images.githubusercontent.com/17103888/197212863-f2fe0517-5718-4a51-a499-6721d165edbc.mov

NevilleS commented:
Thanks @TheAndrewJackson, those toasts are a great improvement. I've pushed a commit to (try) to fix the failing test as well - this should close #1525 as well (updated the description)

TheAndrewJackson commented:
@NevilleS I ran into a small test failure on my machine. It was returning `multimatch/[second].html`  instead of `multimatch/[first].html`. This is because they both generate the same regex pattern and the second file overwrites the first while the route dict is being generated. I did a quick test and tried to build a nextjs website with two dynamic pages like that in the same directory and it returned this error `Error: You cannot use different slug names for the same dynamic path ('id' !== 'test')`. I think it's safe to remove the `multimatch/[second].html`  file from the test data because nextjs prevents builds like that from happening. Removing that file made the test pass.

I'm going to remove the extra test file and if the CI passes I'll merge it.

allisonking commented on clients/privacy-center/common/toast-options.ts:
```diff
@@ -0,0 +1,21 @@
+import {UseToastOptions} from "@chakra-ui/react";
```
surprised prettier didn't make this `{ UseToastOptions }`

allisonking commented on clients/privacy-center/common/toast-options.ts:
```diff
@@ -0,0 +1,21 @@
+import {UseToastOptions} from "@chakra-ui/react";
+
+const BaseToastOptions: UseToastOptions = {
+    position: "top",
+    duration: 5500
+}
```
FYI there's this helper file in admin-ui which I plan to pull into fidesui soon which does similar stuff as this one https://github.com/ethyca/fides/blob/main/clients/admin-ui/src/features/common/toast.tsx

I do think it's nice the way you have the actual error in this file though, should make it easy to import and stay consistent with our copy! I'll have to noodle on it some more, but I might prefer the way you have here with just exporting the ToastOptions instead of what is in admin-ui of exporting a function 👍

allisonking commented on clients/privacy-center/components/modals/VerificationForm.tsx:
```diff
@@ -12,36 +12,36 @@ import {
+import { useLocalStorage } from "~/common/hooks";
 
-import { hostUrl } from "../../constants";
+import { hostUrl } from "~/constants";
```
heh thanks for changing these to `~/`, I do like it a lot more 😄

allisonking commented on clients/privacy-center/components/modals/VerificationForm.tsx:
```diff
@@ -86,13 +91,19 @@ const useVerificationForm = ({
-          handleError(data?.detail);
+          handleError({
+            title: "An error occurred while verifying your identity",
+            error: data?.detail,
```
should we still have the `fallbackMessage` here?

TheAndrewJackson replied:
I think it's okay to leave it out here. The title is very similar to the original `fallbackMessage`. I'm using the error field here for more specific responses from the backend.

The previous version wasn't doing a title and error message at the same time. So the title is acting as a fallback already and the error message is only used if more data is provided from the backend.

allisonking commented on clients/privacy-center/components/modals/consent-request-modal/ConsentRequestForm.tsx:
```diff
@@ -67,33 +75,31 @@ const useConsentRequestForm = ({
-        if (!isVerificationRequired && data.consent_request_id) {
-          setConsentRequestId(data.consent_request_id);
-          successHandler();
+        if (!data.consent_request_id) {
```
woot thanks for doing this! easier to follow now 😄

allisonking commented on clients/privacy-center/components/modals/consent-request-modal/ConsentRequestModal.tsx:
```diff
@@ -91,7 +86,6 @@ export const ConsentRequestModal: React.FC<ConsentRequestModalProps> = ({
       <VerificationForm
         isOpen={isOpen}
         onClose={onClose}
-        setAlert={setAlert}
```
ahhh I love that you were able to clean this up 💯

TheAndrewJackson replied:
It felt good to remove all that code!

allisonking commented on clients/privacy-center/components/modals/privacy-request-modal/PrivacyRequestForm.tsx:
```diff
@@ -66,10 +66,17 @@ const usePrivacyRequestForm = ({
+      }) => {
+        toast({
+          title,
+          description: error,
```
I wonder if it would make sense to do an `error || fallbackErrorMessage` here? so the other places that call this func below don't have to pass in the same "an unhandled error" occurred msg

TheAndrewJackson replied:
Piggy backing off of my comment from another toast related feedback.

> I think it's okay to leave it out here. The title is very similar to the original fallbackMessage. I'm using the error field here for more specific responses from the backend. 
 The previous version wasn't doing a title and error message at the same time. So the title is acting as a fallback already and the error message is only used if more data is provided from the backend.

Since the title is already acting as a fallback message and there aren't useful messages being return from the backend I left the error field blank.


And the titles themselves are very similiar but not quite the same. One says

> error occurred while processing

while the other says

> error occurred while creating

If this is a confusing pattern to use for the toast messages maybe we can interate on it in the future once the toast helper is merged into fidesui?

allisonking commented on clients/privacy-center/pages/_app.tsx:
```diff
@@ -1,6 +1,12 @@
 import type { AppProps } from "next/app";
 import { SessionProvider } from "next-auth/react";
-import { FidesProvider } from "@fidesui/react";
+/*
+ * This import needed to be updated to '@chakra-ui/react' from "@fidesui/react".
+ * Under the hood fidesui is importing from "@chakra-ui/provider" instead "chakra-ui/react".
+ * This causes issues with toasts because it doesn't set up everything required for them.
+ * Solution found here https://github.com/chakra-ui/chakra-ui/issues/5839#issuecomment-1266493711
+ * */
+import { ChakraProvider } from "@chakra-ui/react";
```
great find + comment! it looks like that's how we do it in `admin-ui` too

TheAndrewJackson replied:
Maybe we should update this in `fidesui`? Either that or get rid of the aliasing types altogether. My guess is someone ran into the same problem I did in the `admin-ui` and did exactly what I did.

allisonking commented on src/fides/api/ops/api/v1/endpoints/consent_request_endpoints.py:
```diff
@@ -139,6 +139,27 @@ def consent_request_verify(
+                },
+            ]
+        },
+        HTTP_404_NOT_FOUND: {"detail": "Consent request not found"},
```
I think there also needs to be a `400` error code here which occurs when `subject_identity_verification_required` is turned off?

TheAndrewJackson replied:
Good catch! It's fixed now.

allisonking reviewed (commented):
Looking good! I think there's just one additional docs response that should be added. everything else is just nits

NevilleS reviewed (approved):
LGTM now, thanks for the extra polish based on my QA testing 👍

allisonking reviewed (approved):
🚀
